packages/engine/src/input/systems/ClientInputSystem.ts

```diff
@@ -1,177 +1,212 @@
+  const handleMouseClick = (event: MouseEvent) => {
+    const down = event.type === 'mousedown' || event.type === 'touchstart'
+
+    let button: ButtonTypes = 'PrimaryClick'
+    if (event.button === 1) button = 'AuxiliaryClick'
+    else if (event.button === 2) button = 'SecondaryClick'
+
+    if (down) keyState[button].set(true)
+    else keyState[button].set(none)
+
+    if (down) {
+      const now = Date.now()
+      if (button === 'PrimaryClick') {
+        if (now - lastPrimaryClickDown < 200 && now - lastPrimaryClickDown > 50)
+          keyState['PrimaryDoubleClick'].set(true)
+        lastPrimaryClickDown = now
       }
+
+      if (button === 'AuxiliaryClick') {
+        if (now - lastAuxiliaryClickDown < 200 && now - lastAuxiliaryClickDown > 50)
+          keyState['AuxiliaryDoubleClick'].set(true)
+        lastAuxiliaryClickDown = now
+      }
+
+      if (button === 'SecondaryClick') {
+        if (now - lastSecondaryClickDown < 200 && now - lastSecondaryClickDown > 50)
+          keyState['SecondaryDoubleClick'].set(true)
+        lastSecondaryClickDown = now
+      }
+    } else {
+      if (button === 'PrimaryClick' && keyState['PrimaryDoubleClick'].value) keyState['PrimaryDoubleClick'].set(none)
+      if (button === 'PrimaryClick' && keyState['PrimaryMove'].value) keyState['PrimaryMove'].set(none)
+
+      if (button === 'SecondaryClick' && keyState['SecondaryDoubleClick'].value)
+        keyState['SecondaryDoubleClick'].set(none)
+      if (button === 'SecondaryClick' && keyState['SecondaryMove'].value) keyState['SecondaryMove'].set(none)
+
+      if (button === 'AuxiliaryClick' && keyState['AuxiliaryDoubleClick'].value)
+        keyState['AuxiliaryDoubleClick'].set(none)
+      if (button === 'AuxiliaryClick' && keyState['AuxiliaryMove'].value) keyState['AuxiliaryMove'].set(none)
     }
   }
 
-  // now that we have the data mapped, run the behaviors
-  for (const [key, value] of inputComponent.data) {
-    if (inputComponent.schema.behaviorMap.has(key)) {
-      inputComponent.schema.behaviorMap.get(key)!(entity, key, value)
+  const handleMouseMove = (event: MouseEvent) => {
+    if (keyState['PrimaryClick'].value) keyState['PrimaryMove'].set(true)
+    if (keyState['SecondaryClick'].value) keyState['SecondaryMove'].set(true)
+    if (keyState['AuxiliaryClick'].value) keyState['AuxiliaryMove'].set(true)
+    world.pointerState.position.set(
+      (event.clientX / window.innerWidth) * 2 - 1,
+      (event.clientY / window.innerHeight) * -2 + 1
+    )
+  }
+
+  const handleTouchMove = (event: TouchEvent) => {
+    if (keyState['PrimaryClick'].value) keyState['PrimaryMove'].set(true)
+    if (keyState['SecondaryClick'].value) keyState['SecondaryMove'].set(true)
+    if (keyState['AuxiliaryClick'].value) keyState['AuxiliaryMove'].set(true)
+    const touch = event.touches[0]
+    world.pointerState.position.set(
+      (touch.clientX / window.innerWidth) * 2 - 1,
```

Comment:
I'm hesitant to make this kind of input reactive; this input is likely to change every frame, which means anyone who uses this state in their reactor will schedule reactor updates every frame. This seems like a dangerous pattern. State that changes every frame should really be handled in a system. Reactor updates are much more expensive than systems, and best reserved for reacting to complex discrete state changes.

Reply:
This is not a state reactor, just a simple vector2

Reply:
Got it, misunderstood :)